Clarify interaction between collapsing and multi-line containers (#36791)
mdn/content: files/en-us/web/css/css_flexible_box_layout/mastering_wrapping_of_flex_items/index.md

This behavior is useful if you want to target flex items using JavaScript to show and hide content for example. The example in the specification demonstrates one such pattern.

> In the following live example, the non-wrapping flex container contains a row with three flex items that are set to flex to equal sizes. The third item has multiple lines of content, growing the container. The default for `align-items` is `normal`; for flex items, `normal` behaves as `stretch`, so all the items stretch by default, filling the container's cross-size height.

Author: "the non-wrapping flex container contains a row with three flex items": I wonder if it could be confusing to say it *contains* a row, when really it *is* just a single row, being a single-line flex container.

{{EmbedLiveSample("visibility-collapse")}}

Note that multi-line flex containers rewrap their items after removing collapsed items from rendering. The new space that a collapsed item leaves in the inline direction may cause non-collapsed items to be placed in a different line than if the item were not collapsed. Because each line is laid out like an independent single-line flex container and its composition may change after collapsing, its cross-axis size may change too.

> The following example shows this behavior. The fifth flex item is collapsed, so it occupies zero horizontal space, and so there is enough space to place it on the first row after the fourth item, and so the first row becomes tall enough to fit the three lines of text that the fifth item would have had. Then, if you uncollapse the item (e.g. by removing the `collapse` class), there is no longer enough horizontal space for it on the first row, and it moves to the second. This causes the first row to be only as tall as necessary to fit its new contents (one line) and the second row to be taller instead. Note also that the last flex item is pushed on to a new row, making the entire flex container taller than it was before.

Reviewer: ```suggestion
The following example shows this behavior. The third flex item is collapsed, so it occupies zero space along the main axis (the inline-size is `0`). When collapsed, its strut is on the first row after the fourth item, with the first row being tall enough to fit the three lines of text that the third item would have had. Then, if you uncollapse the item (e.g. by removing the `collapse` class), there is no longer enough horizontal space for the fifth item on the first row, and it moves to the second. This causes the second row to grow to fit the two lines of text of its new member, and the last flex item is pushed onto a new row. With a taller second line and a new third line, the flex container is much taller than it was before.
```

> [!NOTE]
> Use Firefox for the example below as other common browsers treat `collapse` as `hidden`.

```html live-sample___wrapped-visibility-collapse
<div class="box">
  <div>One</div>

>   <div>Two is the width of this sentence.</div>
>   <div class="hide">Three <br />is <br />five <br />lines <br />tall.</div>

Reviewer: ```suggestion
  <div>Two is the width of this sentence.</div>
  <div class="collapse">Three <br />is <br />five <br />lines <br />tall.</div>
```

  <div>Four</div>

>   <div>Five</div>

Reviewer: ```suggestion
  <div>Five<br />Five</div>
```

>   <div>Two</div>
>   <div>Three</div>
>   <div>Four four four four four four four</div>
>   <div class="collapse">Five <br />is <br />taller</div>

Reviewer: ```suggestion
  <div>Two is the width of this sentence.</div>
  <div class="hide">Three <br />is <br />five <br />lines <br/>tall.</div>
  <div>Four</div>
  <div>Five</div>
```
The point of the original is that the first line is as tall as three, even though it is not visible in Firefox. Because it is zero px wide, item 5 is on the first line.
In other browsers, it is hidden, not collapsed, so it is taller, and as wide as the widest line. Because it is wide, item five is on the second line. I think that was the point of the original demo.

The update has a bit too much going on , and may be confusing. We can do two demos if needed; but lets not demonstrate more than one thing in any demo.

Author: Thanks for the feedback. I didn’t think the original was trying to illustrate the difference between Firefox and other browsers, because it specifically said to use Firefox for the demos, and then in a later section said you could change `collapse` to `hidden` to see what it would look like in other browsers. But do you think this example should illustrate that difference?

Otherwise, I interpreted the point of this example as illustrating how a collapsed item affects the cross-axis size of the line containing it and the distribution of other items across lines in a multi-line container. The previous example illustrates the cross-axis size for a single-line container. How would you suggest we split this multi-line example to be clearer?

  border-radius: 5px;
  background-color: rgb(96 139 168 / 0.2);
}
.collapse {
  visibility: collapse;
}
```

```css hidden live-sample___visibility-collapse
p:has(:checked) + div .collapse {
  visibility: visible;
}
```

{{EmbedLiveSample("visibility-collapse")}}

The above was a single-line, non-wrapping flex container with a set size of `600px` so whether the item is visible or collapsed, the width is the same. It is important to understand that while the container retains a strut of the collapsed item's cross-size, the main size is not preserved. Multi-line flex containers rewrap their items after removing collapsed items from rendering. The new space that a collapsed item leaves in the main direction may cause non-collapsed items to be placed in a different line than if the item were not collapsed. Because each line is laid out like an independent single-line flex container and its composition may change after collapsing, its cross-axis size may change too.

> The following example shows this behavior. The third flex item is collapsed, so it occupies zero space along the main axis (the inline-size is `0`). When collapsed, its strut is on the first row after the fourth item, with the first row being tall enough to fit the three lines of text that the third item would have had. Then, if you uncollapse the item (e.g. by removing the `collapse` class), there is no longer enough horizontal space for the fifth item on the first row, and it moves to the second. This causes the second row to grow to fit the two lines of text of its new member, and the last flex item is pushed onto a new row. With a taller second line and a new third line, the flex container is much taller than it was before.

Author: The description does not match the demo (at least for me, on Firefox), but it does if the width of the container is increased from 500px to 520px.

  background-color: rgb(96 139 168 / 0.2);

>   flex: 1 1 auto;

Reviewer: ```suggestion
  flex: 1 1 auto;
  min-width: 50px;
```

Reviewer: in case someone has super small font sizing.

Reviewer 2: Done
